<a href="https://colab.research.google.com/github/AlejitaToloza/Python_Data_Science/blob/main/Utilizando_chatgpt_como_asistente.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# <strong><font color="7b71ff">Análisis de datos con python: Utilizando ChatGPT como asistente</font></strong>


---

<img src="https://raw.githubusercontent.com/alura-cursos/ia-datascience-dataviz/main/imagens/logos/logo_branca_fundo_azul.png" alt="inserir alt">

# **Aula 1 - Conociendo los datos**

 Zoop es una gran minorista que atiende a todas las regiones de México a través de su e-commerce. Es conocida por la amplia variedad de productos que ofrece, buscando satisfacer a todo tipo de público. Posee una serie de centros de distribución (CDs) para la entrega de sus productos, la mayoría de ellos situados en el estado de Mexico. También posee una cadena de supermercados, y vamos a trabajar como científicos de datos para esta prestigiosa empresa.

### **Accesando a los datos**

In [3]:
url = 'https://gist.githubusercontent.com/ahcamachod/24a6235f7cdf255501b5ab4f9b0fd97f/raw/fa329f6d6c1b6b00b3ba457cdc850b53210633ca/zoop_market.json'

> #### <strong><font color="7b71ff" size =6>Prompt:</font></strong>

Vamos a actuar como científicos de datos en Zoop.

Recibimos datos en formato de archivo JSON, estructurados en listas y diccionarios. El archivo está disponible en una URL de Github.

Necesitamos analizar esos datos para ayudar en las decisiones de la empresa. Programamos en Python y conocemos las bibliotecas de análisis de datos Numpy y Pandas. Por lo tanto, prioriza el uso de estas bibliotecas.

Igualmente, infórmanos cómo podemos leer estos datos en formato JSON y para poder visualizarlos en formato de tabla en Python.

In [4]:
import pandas as pd
import numpy as np

# Leer los datos directamente desde la URL
df = pd.read_json(url)

In [5]:
df

index item_identificador tienda_identificador  ventas_totales  \
0         0              FDB08               OUT018       176503.58   
1         1              DRQ35               OUT049       185758.20   
2         2              FDD14               OUT018       165983.94   
3         3              FDY37               OUT045       314923.40   
4         4              FDY59               OUT018        64782.34   
...     ...                ...                  ...             ...   
8545   8545              FDY08               OUT010        28096.76   
8546   8546              FDC41               OUT017       130163.90   
8547   8547              NCQ53               OUT045       614533.40   
8548   8548              FDL46               OUT017       164985.24   
8549   8549              NCN30               OUT046        96541.00   

                                                   item  \
0     {'item_peso': 6.055, 'item_contenido_grasa': '...   
1     {'item_peso': 9.3, 'item_contenido_grasa': 'Ba...   
2     {'item_peso': 20.7, 'item_contenido_grasa': 'B...   
3     {'item_peso': 17.0, 'item_contenido_grasa': 'R...   
4     {'item_peso': 8.195, 'item_contenido_grasa': '...   
...                                                 ...   
8545  {'item_peso': 9.395, 'item_contenido_grasa': '...   
8546  {'item_peso': 15.6, 'item_contenido_grasa': 'B...   
8547  {'item_peso': 17.6, 'item_contenido_grasa': 'B...   
8548  {'item_peso': 20.35, 'item_contenido_grasa': '...   
8549  {'item_peso': 16.35, 'item_contenido_grasa': '...   

                                                 tienda  
0     {'tienda_ano_establecimiento': 2019, 'tienda_t...  
1     {'tienda_ano_establecimiento': 2009, 'tienda_t...  
2     {'tienda_ano_establecimiento': 2019, 'tienda_t...  
3     {'tienda_ano_establecimiento': 2012, 'tienda_t...  
4     {'tienda_ano_establecimiento': 2019, 'tienda_t...  
...                                                 ...  
8545  {'tienda_ano_establecimiento': 2008, 'tienda_t...  
8546  {'tienda_ano_establecimiento': 2017, 'tienda_t...  
8547  {'tienda_ano_establecimiento': 2012, 'tienda_t...  
8548  {'tienda_ano_establecimiento': 2017, 'tienda_t...  
8549  {'tienda_ano_establecimiento': 2007, 'tienda_t...  

[8550 rows x 6 columns]

> #### <strong><font color="7b71ff" size =6>Prompt:</font></strong>


El DataFrame "df" posee las columnas "item_identificador", "tienda_identificador", "ventas_totales", "item" y "tienda".

Las columnas "item" y "tienda" contienen diccionarios anidados dentro de cada una de las filas. ¿Cómo puedo transformar los datos de los diccionarios y columnas para hacerlos más accesibles en mi DataFrame?


In [26]:
# Extraer datos de la columna 'item'
if 'item' in df.columns:
    # Convertir cada diccionario en una fila
    item_df = pd.json_normalize(df['item'])
    item_df.columns = [f'item_{col}' for col in item_df.columns]  # Renombrar columnas

    # Concatenar con el DataFrame original
    df = pd.concat([df.drop('item', axis=1), item_df], axis=1)

# Mostrar el DataFrame resultante
print(df.head())

# El resultado no fue óptimo intentando con normalize, la información contenida en los diccionarios item y tienda no se podían extraer...

   index item_identificador tienda_identificador  ventas_totales  \
0      0              FDB08               OUT018       176503.58   
1      1              DRQ35               OUT049       185758.20   
2      2              FDD14               OUT018       165983.94   
3      3              FDY37               OUT045       314923.40   
4      4              FDY59               OUT018        64782.34   

                                              tienda  \
0  {'tienda_ano_establecimiento': 2019, 'tienda_t...   
1  {'tienda_ano_establecimiento': 2009, 'tienda_t...   
2  {'tienda_ano_establecimiento': 2019, 'tienda_t...   
3  {'tienda_ano_establecimiento': 2012, 'tienda_t...   
4  {'tienda_ano_establecimiento': 2019, 'tienda_t...   

                                              item_0  
0  {'item_peso': 6.055, 'item_contenido_grasa': '...  
1  {'item_peso': 9.3, 'item_contenido_grasa': 'Ba...  
2  {'item_peso': 20.7, 'item_contenido_grasa': 'B...  
3  {'item_peso': 17.0, 'item_conte

In [31]:
# Se recreó todo el proceso nuevamente esta vez aplicando funciones diferentes como eval(x) y pd.Series
# Uso de eval: Se utilizó eval(x) para convertir la cadena que representa un diccionario en un diccionario real.
# Esto solo debe hacerse si estás seguro de que los datos son seguros y no contienen código malicioso.
# Se aplicó pd.Series: Para convertir cada diccionario en una fila de un nuevo DataFrame.
# Renombrar columnas: Se renombraron las columnas para diferenciarlas.
# Concatenar: Se combinó el DataFrame original (sin las columnas "item" y "tienda") con los DataFrames expandidos.

import pandas as pd

# Leer el archivo JSON
url = 'https://gist.githubusercontent.com/ahcamachod/24a6235f7cdf255501b5ab4f9b0fd97f/raw/fa329f6d6c1b6b00b3ba457cdc850b53210633ca/zoop_market.json'
df = pd.read_json(url)

# Extraer datos de la columna 'item'
if 'item' in df.columns:
    # Convertir cada diccionario en una fila
    item_df = df['item'].apply(lambda x: eval(x) if isinstance(x, str) else x).apply(pd.Series)
    item_df.columns = [f'item_{col}' for col in item_df.columns]  # Renombrar columnas

# Extraer datos de la columna 'tienda'
if 'tienda' in df.columns:
    # Convertir cada diccionario en una fila
    tienda_df = df['tienda'].apply(lambda x: eval(x) if isinstance(x, str) else x).apply(pd.Series)
    tienda_df.columns = [f'tienda_{col}' for col in tienda_df.columns]  # Renombrar columnas

# Concatenar los DataFrames
df = pd.concat([df.drop(['item', 'tienda'], axis=1), item_df, tienda_df], axis=1)

# Mostrar el DataFrame resultante
print(df)


      index item_identificador tienda_identificador  ventas_totales  \
0         0              FDB08               OUT018       176503.58   
1         1              DRQ35               OUT049       185758.20   
2         2              FDD14               OUT018       165983.94   
3         3              FDY37               OUT045       314923.40   
4         4              FDY59               OUT018        64782.34   
...     ...                ...                  ...             ...   
8545   8545              FDY08               OUT010        28096.76   
8546   8546              FDC41               OUT017       130163.90   
8547   8547              NCQ53               OUT045       614533.40   
8548   8548              FDL46               OUT017       164985.24   
8549   8549              NCN30               OUT046        96541.00   

      item_item_peso item_item_contenido_grasa  item_item_visibilidad  \
0              6.055   Bajo Contenido de Grasa               0.031230   
1

In [30]:
tienda_df
item_df

item_item_peso item_item_contenido_grasa  item_item_visibilidad  \
0              6.055   Bajo Contenido de Grasa               0.031230   
1              9.300   Bajo Contenido de Grasa               0.042357   
2             20.700   Bajo Contenido de Grasa               0.170500   
3             17.000                   Regular               0.026623   
4              8.195   Bajo Contenido de Grasa               0.000000   
...              ...                       ...                    ...   
8545           9.395                   Regular               0.286345   
8546          15.600   Bajo Contenido de Grasa               0.117575   
8547          17.600   Bajo Contenido de Grasa               0.018944   
8548          20.350   bajo contenido de grasa               0.054363   
8549          16.350                       BCG               0.016993   

            item_item_tipo  item_item_precio item_item_cantidad_venta  
0       Frutas y Vegetales            160.36                     None  
1      Bebidas Alcohólicas            123.24                     None  
2                Enlatados            184.13                     None  
3                Enlatados            144.25                     None  
4                 Dulcería             93.15                     None  
...                    ...               ...                      ...  
8545    Frutas y Vegetales            139.18                     None  
8546  Alimentos Congelados             75.67                     None  
8547             Abarrotes            237.36                     None  
8548                Snacks            117.95                     None  
8549              Cereales             95.74                     None  

[8550 rows x 6 columns]

# **Clase 2 - Exploración de los dados**

> #### <strong><font color="7b71ff" size =6>Prompt:</font></strong>

El DataFrame "df" contiene datos referentes a las ventas de diversos productos en varias tiendas del supermercado. Se proporcionan los IDs de productos y los IDs de tiendas que comercializaron esos productos, el total recaudado en ventas, así como descripciones sobre el producto vendido y la tienda que lo vendió.

Estamos iniciando la etapa de exploración de datos, en la cual deseamos conocer los valores informados e identificar si existen valores ausentes, duplicados y otras inconsistencias.

Construye un código para ejecutar la etapa de exploración de datos usando Python, comentando el paso a paso de manera clara y objetiva.

No es necesario utilizar "print()" para mostrar el resultado de las consultas.

> #### <strong><font color="7b71ff" size =6>Prompt:</font></strong>

El DataFrame "df" contiene datos relacionados con las ventas durante un trimestre de un supermercado. Deseo utilizarlo para verificar el desempeño de esas ventas.

Durante la exploración del "df", identificamos algunas inconsistencias, como:

* Filas duplicadas,
* Valores nulos y
* Datos categóricos que fueron escritos de manera no estandarizada, dando la idea de que existen más categorías de las que realmente existen.

¿Qué debo hacer para garantizar que esas inconsistencias que encontré no afecten mi análisis?

# **Clase 3 - Limpieza y tratamiento de los datos**

> #### <strong><font color="7b71ff" size =6>Prompt:</font></strong>


Durante la exploración del DataFrame "df", identificamos 27 filas duplicadas usando "df.duplicated().sum()".

El objetivo es analizar los datos en "df" para medir el desempeño de una empresa durante un trimestre de ventas.

¿Qué debemos hacer con los datos duplicados para garantizar que no haya impacto negativo en el análisis?

> #### <strong><font color="7b71ff" size =6>Prompt:</font></strong>


Identificamos tres columnas en "df" que presentan valores nulos:

* La columna "item_cantidad_venta" tiene 8550 filas con valores nulos.
* La columna "item_peso" tiene 1469 filas con valores nulos.
* La columna "tienda_tamano" tiene 2417 filas con valores nulos.

Hubo un gran número de ventas de artículos durante el trimestre, y no hay justificación para la ausencia de datos en la columna "item_cantidad_venta", lo que la convierte en una columna prescindible.

La columna "item_peso" contiene valores del tipo float, representando el peso en gramos de los productos.

La columna "tienda_tamano" tiene tres categorías de tamaño para supermercados.

¿Cómo debo manejar estos datos nulos para garantizar que mi análisis no se vea afectado?

> #### <strong><font color="7b71ff" size =6>Prompt:</font></strong>

La columna "item_conteudo_grasa" del DataFrame "df" presenta 5 valores únicos dados por: 'Bajo Contenido de Grasa', 'Regular', 'BTG', 'reg' y 'bajo contenido de grasa'.

Los valores únicos de la columna "item_contenido_grasa" deberían ser solo 'Bajo Contenido de Grasa' y 'Regular', pero fueron insertados sin estandarización, lo que resultó en más categorías de las deseadas.

¿Cómo puedo ajustar los datos de la columna "item_contenido_grasa" para mantener solo las dos categorías estándar?

# **Clase 4 - Visualización de datos**

> #### <strong><font color="7b71ff" size =6>Prompt:</font></strong>

Los datos del DataFrame "df" están limpios y tratados. Ahora, deseo iniciar la etapa de construcción de gráficos para crear visualizaciones que ofrezcan un buen soporte al análisis del desempeño de ventas de la empresa.

Las columnas disponibles en "df" son las siguientes:
- "item_identificador": Número único asignado a cada ítem
- "tienda_identificador": Nombre único de la tienda
- "ventas_totales": Valor obtenido en ventas totales
- "item_peso": Peso del ítem en gramos
- "item_contenido_grasa": Contenido de grasa del ítem
- "item_visibilidad": Valor de posicionamiento de cada ítem
- "item_tipo": Tipo de utilidad del ítem
- "item_precio": Precio unitario del ítem
- "tienda_ano_establecimiento": Año de Establecimiento de la tienda
- "tienda_tamano": Tamaño de la tienda
- "tienda_tipo_ubicacion": Tipo de ubicación de la tienda
- "tienda_tipo": Tipo de tienda

El objetivo es encontrar patrones o información que se destaquen cuando nos referimos al total obtenido en ventas especificado en la columna "ventas_totales".

Con base en la información proporcionada, ¿qué tipos de visualizaciones serían útiles para alcanzar el objetivo de analizar el desempeño de ventas?

> #### <strong><font color="7b71ff" size =6>Prompt:</font></strong>

El DataFrame "df" contiene la columna "ventas_totales", que informa el valor en reales obtenido en ventas de productos por varias tiendas.

Además, en "df", tenemos la columna "tienda_tamano" con 4 categorías de tamaño para tiendas: 'Grande', 'Medio', 'Pequeño' y 'No registrado'.

Deseo crear una visualización que permita identificar el total obtenido en ventas para las cuatro categorías de tamaño de tienda y representarlas en un gráfico de barras.

Los datos en la columna "ventas_totales" necesitan ser agrupados por las categorías de "tienda_tamano".

La figura del gráfico debe seguir la identidad visual de la tienda, con el fondo de la figura y del gráfico en color "#ebebeb" y las barras en color "#e23155".

Además, es necesario mostrar el total de las ventas en millones encima de cada columna, con el texto de esta información en color "#19325c" y la estructura "MXN $ valor M".

Utiliza las bibliotecas de visualización de datos de Python para crear esta visualización.

> #### <strong><font color="7b71ff" size =6>Prompt:</font></strong>


El DataFrame "df" contiene la columna "ventas_totales", que informa el valor en reales obtenido en ventas de productos por varias tiendas.

Además, en "df", tenemos la columna "tienda_tipo" con 4 categorías de tipo para tiendas: 'Supermercado Tipo 1', 'Supermercado Tipo 2', 'Mercado', 'Supermercado Tipo 3'.

Deseo crear una visualización que permita identificar el total obtenido en ventas para las cuatro categorías de tipo de tienda y representarlas en un gráfico de barras.

Los datos en la columna "ventas_totales" deben ser agrupados por las categorías de "tienda_tipo".

La figura de la gráfica debe seguir la identidad visual de la tienda, con el fondo de la figura y de la gráfica en color "#ebebeb" y las barras en color "#e23155".

Además, es necesario mostrar el total de las ventas en millones sobre cada columna, con el texto de esa información en color "#19325c" y la estructura "MXN$ valor M".

Utiliza las bibliotecas de visualización de datos de Python para crear esta visualización.

> #### <strong><font color="7b71ff" size =6>Prompt:</font></strong>

El DataFrame "df" contiene la columna "ventas_totales", que informa el valor en reales obtenido en ventas de productos por varias tiendas.

Además, en "df", tenemos la columna "item_tipo" con 16 categorías de productos.

Deseo crear una visualización que permita identificar el total obtenido en ventas por categorías de productos en un gráfico de barras horizontales.

Los datos en la columna "ventas_totales" deben agruparse por las categorías de "item_tipo".

La figura de la representación debe seguir la identidad visual de la tienda, con el fondo de la figura y del gráfico en el color "#ebebeb".

Las barras deben estar en el color "#e23155".

Los valores de las barras deben estar ordenados para mantener un aspecto de embudo.

Además, es necesario mostrar el total de las ventas en millones dentro o al lado de cada barra, con el texto de esa información en el color "#19325c" y en la estructura "MXN$ valor M".

No muestres la cuadrícula en el gráfico.

Utiliza las bibliotecas de visualización de datos de Python para crear esta visualización.